iceoryx_posh/source/popo/trigger.cpp

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2020 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iceoryx_posh/popo/trigger.hpp"
+#include "iceoryx_posh/internal/popo/building_blocks/condition_variable_signaler.hpp"
+
+namespace iox
+{
+namespace popo
+{
+std::atomic<uint64_t> Trigger::uniqueIdCounter{0U};
+
+Trigger::~Trigger()
+{
+    reset();
+}
+
+bool Trigger::hasTriggered() const noexcept
+{
+    return (isValid()) ? m_hasTriggeredCallback().value() : false;
+}
+
+void Trigger::reset() noexcept
+{
+    if (!isValid())
+    {
+        return;
+    }
+
+    if (m_resetCallback)
+    {
+        m_resetCallback(m_uniqueId);
+    }
+
+    invalidate();
+}
+
+TriggerState Trigger::getTriggerState() const noexcept
+{
+    return m_triggerState;
+}
+
+void Trigger::invalidate() noexcept
+{
+    m_hasTriggeredCallback = cxx::ConstMethodCallback<bool>();
+}
+
+Trigger::operator bool() const noexcept
+{
+    return isValid();
+}
+
+bool Trigger::isValid() const noexcept
+{
+    return static_cast<bool>(m_hasTriggeredCallback);
+}
+
+bool Trigger::isLogicalEqualTo(const Trigger& rhs) const noexcept
+{
+    return (isValid() && rhs.isValid() && m_triggerState.m_origin == rhs.m_triggerState.m_origin
+            && m_hasTriggeredCallback == rhs.m_hasTriggeredCallback
+            && m_triggerState.m_triggerId == rhs.m_triggerState.m_triggerId);
+}
+
+Trigger::Trigger(Trigger&& rhs) noexcept
+{
+    *this = std::move(rhs);
```

Comment:
Works, but usually it is better to implement operator= in terms of the move ctor for efficiency reasons, but this requires a swap for the operator=. Alternatively slight code duplication using the initializer list to directly initialize the members.

The object is default constructed here and then moved to, wasting the default constructor effort. The object is somewhat nontrivial I would think by looking at the members.

Reply:
This approach is safe and not so error prone as if you would implement it twice. Furthermore, we use the move ctor not very often. The usual way to work with Triggers and a Waitset is
1. create waitset
2. attach classes (here is the move involved)
3. while loop and handle events (99.9 % of the application time is spend in this loop)
4. detach classes